### PR TITLE
[WebGPU] Invalid limits do not result in device creation failure

### PR DESCRIPTION
**Pull request by mwyrzykowski**
#### 403cf88c8df73ed92e7626dd3c1e5511d9762b03
<pre>
[WebGPU] Invalid limits do not result in device creation failure
<a href="https://bugs.webkit.org/show_bug.cgi?id=252700">https://bugs.webkit.org/show_bug.cgi?id=252700</a>
&lt;radar://105751730&gt;

Reviewed by Myles C. Maxfield.

The specification says we should reject the promise of
GPUAdapter.requestDevice when the specified limits are invalid.

Achieve this by returning null on failure from requestDevice
and rejecting the promise in the GPUAdapter layer.

Also update WebGPU.h as maxColorAttachmentBytesPerSample was
missing from WGPULimits.

* Source/WebCore/Modules/WebGPU/GPUAdapter.cpp:
(WebCore::GPUAdapter::requestDevice):
* Source/WebCore/Modules/WebGPU/GPUSupportedLimits.cpp:
(WebCore::GPUSupportedLimits::maxBindingsPerBindGroup const):
(WebCore::GPUSupportedLimits::maxBufferSize const):
(WebCore::GPUSupportedLimits::maxInterStageShaderVariables const):
(WebCore::GPUSupportedLimits::maxColorAttachments const):
(WebCore::GPUSupportedLimits::maxColorAttachmentBytesPerSample const):
* Source/WebCore/Modules/WebGPU/GPUSupportedLimits.h:
* Source/WebCore/Modules/WebGPU/GPUSupportedLimits.idl:
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUAdapterImpl.cpp:
(PAL::WebGPU::supportedFeatures):
(PAL::WebGPU::supportedLimits):
(PAL::WebGPU::AdapterImpl::requestDevice):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUAdapterImpl.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUAdapter.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUSupportedLimits.h:
* Source/WebGPU/WGSL/Overload.cpp:
(WGSL::OverloadResolver::materialize const):
* Source/WebGPU/WebGPU/Adapter.mm:
(WebGPU::Adapter::requestInvalidDevice):
* Source/WebGPU/WebGPU/Buffer.mm:
(wgpuBufferGetMapState):
* Source/WebGPU/WebGPU/ComputePipeline.mm:
(WebGPU::Device::createComputePipelineAsync):
* Source/WebGPU/WebGPU/HardwareCapabilities.mm:
(WebGPU::apple3):
(WebGPU::apple4):
(WebGPU::apple5):
(WebGPU::apple6):
(WebGPU::apple7):
(WebGPU::mac2):
(WebGPU::mergeLimits):
(WebGPU::anyLimitIsBetterThan):
(WebGPU::defaultLimits):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::Device::createRenderPipelineAsync):
* Source/WebGPU/WebGPU/ShaderModule.mm:
(WebGPU::convertMessages):
* Source/WebGPU/WebGPU/WebGPU.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteAdapter.cpp:
(WebKit::RemoteAdapter::requestDevice):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.cpp:
(WebKit::RemoteGPU::requestAdapter):
* Source/WebKit/Shared/WebGPU/WebGPUSupportedLimits.cpp:
(WebKit::WebGPU::ConvertToBackingContext::convertToBacking):
(WebKit::WebGPU::ConvertFromBackingContext::convertFromBacking):
* Source/WebKit/Shared/WebGPU/WebGPUSupportedLimits.h:
* Source/WebKit/Shared/WebGPU/WebGPUSupportedLimits.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteAdapterProxy.cpp:
(WebKit::WebGPU::RemoteAdapterProxy::requestDevice):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteAdapterProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteGPUProxy.cpp:
(WebKit::RemoteGPUProxy::requestAdapter):

Canonical link: <a href="https://commits.webkit.org/260777@main">https://commits.webkit.org/260777@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/9a17dfdac1086b5dbd5180306587f8d89a3e4cf3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109399 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18523 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/42172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/87/builds/907 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/118574 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113282 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/19997 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/9730 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101674 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115155 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/19997 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/42172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/43094 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/76/builds/19997 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/42172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/84846 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/11276 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/42172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/11942 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/9730 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/17309 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/42172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7457 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13664 "Built successfully") | | | 
| | | | |

### CODE REVIEW

File: Source/WebCore/Modules/WebGPU/GPUSupportedLimits.cpp [hunk 1]
```diff
@@ -53,6 +53,11 @@ uint32_t GPUSupportedLimits::maxBindGroups() const
     return m_backing->maxBindGroups();
 }
 
+uint32_t GPUSupportedLimits::maxBindingsPerBindGroup() const
+{
+    return m_backing->maxBindingsPerBindGroup();
```
Comment (litherum) [inline Source/WebCore/Modules/WebGPU/GPUSupportedLimits.cpp:58]: Usually this data flow is backwards; information doesn't usually flow from the backing to the front-ing. Are you sure you want to ask the backing, instead of just having the information be owned right here in `GPUSupportedLimits` and set at creation time?

File: Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUAdapterImpl.cpp [hunk 2]
```diff
@@ -144,7 +152,7 @@ AdapterImpl::~AdapterImpl()
     wgpuAdapterRelease(m_backing);
 }
 
-void AdapterImpl::requestDevice(const DeviceDescriptor& descriptor, CompletionHandler<void(Ref<Device>&&)>&& callback)
+void AdapterImpl::requestDevice(const DeviceDescriptor& descriptor, CompletionHandler<void(RefPtr<Device>&&)>&& callback)
```
Comment (litherum) [inline Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUAdapterImpl.cpp:191]: How confident are you that we shouldn't be returning an invalid device instead of nullptr?

File: Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUAdapterImpl.cpp [hunk 3]
```diff
@@ -171,24 +179,23 @@ void AdapterImpl::requestDevice(const DeviceDescriptor& descriptor, CompletionHa
...
         .maxComputeWorkgroupsPerDimension =    65535,
     };
 
-    auto requestInvalidDevice = [this, &callback]() {
-        wgpuAdapterRequestInvalidDeviceWithBlock(m_backing, makeBlockPtr([protectedThis = Ref { *this }, convertToBackingContext = m_convertToBackingContext.copyRef(), callback = WTFMove(callback)](WGPUDevice device) mutable {
-            callback(DeviceImpl::create(device, Ref { protectedThis->features() }, Ref { protectedThis->limits() }, convertToBackingContext));
-        }).get());
+    auto requestInvalidDevice = [&callback]() {
+        callback(nullptr);
     };
```
Comment (litherum) [inline Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUAdapterImpl.cpp]: Is this really worth it? If it were me, I'd just inline this lambda at call sites.

File: Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUAdapterImpl.cpp [hunk 4]
```diff
@@ -224,18 +231,23 @@ void AdapterImpl::requestDevice(const DeviceDescriptor& descriptor, CompletionHa
...
         SET_VALUE_32(maxComputeWorkgroupSizeY)
         SET_VALUE_32(maxComputeWorkgroupSizeZ)
         SET_VALUE_32(maxComputeWorkgroupsPerDimension)
+        else {
+            requestInvalidDevice();
+            return;
+        }
 
 #undef SET_VALUE_32
 #undef SET_VALUE_64
```
Comment (litherum) [inline Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUAdapterImpl.cpp]: There has got to be a better way of doing this without the preprocessor.

File: Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUSupportedLimits.h [hunk 5]
```diff
@@ -25,6 +25,7 @@
 
 #pragma once
 
+#include <WebGPU/WebGPU.h>
```
Comment (litherum) [inline Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUSupportedLimits.h]: Nope. Layering violation. r-

File: Source/WebGPU/WGSL/Overload.cpp [hunk 6]
```diff
@@ -143,6 +143,9 @@ Type* OverloadResolver::materialize(const AbstractType& abstractType) const
         },
         [&](TypeVariable variable) -> Type* {
             auto* resolvedType = resolve(variable);
+            if (!resolvedType)
+                return nullptr;
+
```
Comment (litherum) [inline Source/WebGPU/WGSL/Overload.cpp]: This looks misplaced. Are you sure this shouldn't be in a separate patch?

File: Source/WebGPU/WebGPU/Adapter.mm [hunk 7]
```diff
@@ -130,9 +130,7 @@
 
 void Adapter::requestInvalidDevice(CompletionHandler<void(Ref<Device>&&)>&& callback)
 {
-    instance().scheduleWork([strongThis = Ref { *this }, callback = WTFMove(callback)]() mutable {
-        callback(Device::createInvalid(strongThis));
-    });
+    callback(Device::createInvalid(*this));
```
Comment (litherum) [inline Source/WebGPU/WebGPU/Adapter.mm]: I think we've already discussed this ... Device creation is supposed to be async.

File: Source/WebGPU/WebGPU/Buffer.mm [hunk 8]
```diff
@@ -356,6 +356,22 @@ void wgpuBufferDestroy(WGPUBuffer buffer)
     return WebGPU::fromAPI(buffer).getConstMappedRange(offset, size);
 }
 
+WGPUBufferMapState wgpuBufferGetMapState(WGPUBuffer buffer)
+{
+    switch (WebGPU::fromAPI(buffer).state()) {
+    case WebGPU::Buffer::State::Mapped:
```
Comment (litherum) [inline Source/WebGPU/WebGPU/Buffer.mm:362]: Should we just get rid of our own enum in favor of the WGPU one?

File: Source/WebKit/GPUProcess/graphics/WebGPU/RemoteAdapter.cpp [hunk 9]
```diff
@@ -66,7 +66,13 @@ void RemoteAdapter::requestDevice(const WebGPU::DeviceDescriptor& descriptor, We
         return;
     }
 
-    m_backing->requestDevice(*convertedDescriptor, [callback = WTFMove(callback), objectHeap = Ref { m_objectHeap }, streamConnection = m_streamConnection.copyRef(), identifier, queueIdentifier] (Ref<PAL::WebGPU::Device>&& device) mutable {
+    m_backing->requestDevice(*convertedDescriptor, [callback = WTFMove(callback), objectHeap = Ref { m_objectHeap }, streamConnection = m_streamConnection.copyRef(), identifier, queueIdentifier] (RefPtr<PAL::WebGPU::Device>&& devicePtr) mutable {
+        if (!devicePtr.get()) {
```
Comment (litherum) [inline Source/WebKit/GPUProcess/graphics/WebGPU/RemoteAdapter.cpp:70]: Ditto about invalid device vs nullptr. Generally, things shouldn't be null, but should be invalid instead. Are we super confident that null is actually correct here?

File: Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteAdapterProxy.cpp [hunk 10]
```diff
@@ -46,7 +46,7 @@ RemoteAdapterProxy::~RemoteAdapterProxy()
 {
 }
 
-void RemoteAdapterProxy::requestDevice(const PAL::WebGPU::DeviceDescriptor& descriptor, CompletionHandler<void(Ref<PAL::WebGPU::Device>&&)>&& callback)
+void RemoteAdapterProxy::requestDevice(const PAL::WebGPU::DeviceDescriptor& descriptor, CompletionHandler<void(RefPtr<PAL::WebGPU::Device>&&)>&& callback)
```
Comment (litherum) [inline Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteAdapterProxy.cpp:49]: Ditto.

File: Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteAdapterProxy.cpp [hunk 11]
```diff
@@ -60,13 +60,19 @@ void RemoteAdapterProxy::requestDevice(const PAL::WebGPU::DeviceDescriptor& desc
         return;
 
     auto [supportedFeatures, supportedLimits] = sendResult.takeReply();
+    if (!supportedLimits.maxTextureDimension2D) {
+        callback(nullptr);
+        return;
+    }
```
Comment (litherum) [inline Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteAdapterProxy.cpp:66]: Why is this limit special? This looks like it shouldn't be here.

File: Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteAdapterProxy.h [hunk 12]
```diff
@@ -76,8 +76,13 @@ class RemoteAdapterProxy final : public PAL::WebGPU::Adapter {
     {
         return root().streamClientConnection().sendSync(WTFMove(message), backing(), defaultSendTimeout);
     }
+    template<typename T, typename C>
+    WARN_UNUSED_RETURN IPC::StreamClientConnection::AsyncReplyID sendWithAsyncReply(T&& message, C&& completionHandler)
```
Comment (litherum) [inline Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteAdapterProxy.h]: Can we just add this everywhere that needs it in a single standalone patch, instead of adding it lazily? Every async call will need it.

File: Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteAdapterProxy.h [hunk 13]
```diff
@@ -76,8 +76,13 @@ class RemoteAdapterProxy final : public PAL::WebGPU::Adapter {
...
+    {
+        return root().streamClientConnection().sendWithAsyncReply(WTFMove(message), completionHandler, backing(), defaultSendTimeout);
+    }
 
-    void requestDevice(const PAL::WebGPU::DeviceDescriptor&, CompletionHandler<void(Ref<PAL::WebGPU::Device>&&)>&&) final;
+    void requestDevice(const PAL::WebGPU::DeviceDescriptor&, CompletionHandler<void(RefPtr<PAL::WebGPU::Device>&&)>&&) final;
```
Comment (litherum) [inline Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteAdapterProxy.h:80]: Ditto from above about nullptr.

**Review by litherum — APPROVED** [review]
I'm saying r+ but please don't land until you've fixed the layering violation.

File: Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUAdapterImpl.cpp (hunk 2 above)
Comment (mwyrzykowski) [reply Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUAdapterImpl.cpp:191]: The specification says to reject the promise. So I don't think there is anywhere to return the invalid device.

File: Source/WebGPU/WebGPU/Adapter.mm (hunk 7 above)
Comment (mwyrzykowski) [reply Source/WebGPU/WebGPU/Adapter.mm]: This code is not even used actually so I will remove it.

File: Source/WebGPU/WGSL/Overload.cpp (hunk 6 above)
Comment (mwyrzykowski) [reply Source/WebGPU/WGSL/Overload.cpp]: Thanks, that should not be in this patch

File: Source/WebGPU/WebGPU/Buffer.mm (hunk 8 above)
Comment (mwyrzykowski) [reply Source/WebGPU/WebGPU/Buffer.mm:362]: Our own enum has additional information (MappedAtCreation and Destroy), I'm not sure if that is useful.

File: Source/WebKit/GPUProcess/graphics/WebGPU/RemoteAdapter.cpp (hunk 9 above)
Comment (mwyrzykowski) [reply Source/WebKit/GPUProcess/graphics/WebGPU/RemoteAdapter.cpp:70]: Fairly certain null is correct. The device is never returned since the promise is rejected. If we created an invalid deice, what would we do with it? We don't return it via the promise

File: Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteAdapterProxy.cpp (hunk 11 above)
Comment (mwyrzykowski) [reply Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteAdapterProxy.cpp:66]: This limit is not special but it is to signify the device is not created via returning an empty value. We could check the entire structure, but it is not needed. If 0 is returned for something as basic as max 2d texture size, the device is not valid.

File: Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteAdapterProxy.h (hunk 12 above)
Comment (mwyrzykowski) [reply Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteAdapterProxy.h]: Yes good idea. I removed it from this change since it is not used

File: Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteAdapterProxy.cpp (hunk 11 above)
Comment (mwyrzykowski) [reply Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteAdapterProxy.cpp:66]: I tried returning an optional<tuple<a, b>> from the IPC message instead of the tuple<a,b> that is currently returned but that became a lot of unnecessary changes and this approach is much simpler

File: Source/WebCore/Modules/WebGPU/GPUSupportedLimits.cpp (hunk 1 above)
Comment (mwyrzykowski) [reply Source/WebCore/Modules/WebGPU/GPUSupportedLimits.cpp:58]: I think the change is outside the scope of this PR however isn't this the pattern we follow with other GPU* classes?

For instance, every ::label implementation looks like:
```
String GPU[Buffer]::label() const
{
    return m_backing->label();
}
```

We can certainly change it and I think having the information stored directly in GPUSupportedLimits as suggested would be better.

File: Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUAdapterImpl.cpp (hunk 4 above)
Comment (mwyrzykowski) [reply Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUAdapterImpl.cpp]: We could use templates to avoid have the _32 and _64 variants. Not immediately obvious how to get rid of the string conversion without some reflection information
